### PR TITLE
Support of complex linear systems

### PR DESCRIPTION
**Pull request by amontoison**
@dpo, @tmigot, @geoffroyleconte
I updated the Krylov methods to support complex numbers.
Can you give your opinion before I "finalize" the update of each method and add the tests ?
Some methods only require a few `abs()` and `conj()` to handle complex linear systems.

I defined a new type :
```
"""
    FloatOrComplex{T}
Union type of `T` and `Complex{T}` where T is an [`AbstractFloat`](@ref).
"""
const FloatOrComplex{T} = Union{T, Complex{T}} where T <: AbstractFloat
```
and  I replaced each method
 ```
xyz(A, b :: AbstractVector{T}; kwargs...) where T <: AbstractFloat
```
by
```
xyz(A, b :: AbstractVector{FC}; kwargs...) where FC <: FloatOrComplex
```

The in-place methods and `KrylovSolvers` are now parametrized by 3 parameters `{T,FC,S}` with `{T <: AbstractFloat, FC <: FloatOrComplex{T}, S <: DenseVector{FC}}`.

@abelsiqueira 
If should be still possible to do
```
KrylovSolver{T,FC,S} <: AbstractJSOSolver{T, S}
```
if you want to do a generic solver. :smiley:

### CODE REVIEW

**Comment by dpo** [comment]
Please update 1 method per PR with tests to begin with and first determine the changes required for Lanczos-based method, then GK, then etc.

**Comment by amontoison** [comment]
@JSOBot runbenchmarks

**Comment by amontoison** [comment]
@JSOBot runbenchmarks

**Comment by MonssafToukal** [comment]
@JSOBot runbenchmarks

File: src/krylov_utils.jl [hunk 1]
```diff
@@ -1,3 +1,9 @@
+"""
+    FloatOrComplex{T}
+Union type of `T` and `Complex{T}` where T is an [`AbstractFloat`](@ref).
+"""
+const FloatOrComplex{T} = Union{T, Complex{T}} where T <: AbstractFloat
+
```
Comment (amontoison) [inline src/krylov_utils.jl:6]: @tmigot

**Comment by amontoison** [comment]
I will merge this PR in the branch `complex` of JSO/Krylov.jl and after update one by one the Krylov methods.
This PR only updates the Krylov solvers, docstrings and headers.

**Pull request merged**